swaps/swaps_service.go

```diff
@@ -125,46 +144,117 @@ func (svc *swapsService) EnableAutoSwaps(ctx context.Context, lnClient lnclient.
 		}
 	}()
 
-	svc.cancelFn = cancelFn
+	svc.swapOutCancelFn = cancelFn
 
 	return nil
 }
 
-func (svc *swapsService) StopAutoSwaps() {
-	if svc.cancelFn != nil {
-		logger.Logger.Info("Stopping swap service...")
-		svc.cancelFn()
-		logger.Logger.Info("swap service stopped")
+func (svc *swapsService) EnableAutoSwapIn(ctx context.Context, lnClient lnclient.LNClient) error {
+	// stop any existing swap in process
+	svc.StopAutoSwap(true, false)
+
+	// TODO: change threshold keys
+	ctx, cancelFn := context.WithCancel(ctx)
+	balanceThresholdStr, _ := svc.cfg.Get(config.AutoSwapInBalanceThresholdKey, "")
+	amountStr, _ := svc.cfg.Get(config.AutoSwapInAmountKey, "")
+
+	if balanceThresholdStr == "" || amountStr == "" {
+		cancelFn()
+		logger.Logger.Info("Auto swap not configured")
+		return nil
+	}
+
+	balanceThreshold, err := strconv.ParseUint(balanceThresholdStr, 10, 64)
+	if err != nil {
+		cancelFn()
+		return errors.New("invalid auto swap configuration")
+	}
+
+	amount, err := strconv.ParseUint(amountStr, 10, 64)
+	if err != nil {
+		cancelFn()
+		return errors.New("invalid auto swap configuration")
 	}
+
+	logger.Logger.Info("Starting auto swap workflow")
+
+	go func() {
+		ticker := time.NewTicker(1 * time.Hour)
+		for {
+			select {
+			case <-ticker.C:
+				logger.Logger.Debug("Checking to see if we can swap")
+				balance, err := lnClient.GetBalances(ctx, false)
+				if err != nil {
+					logger.Logger.WithError(err).Error("Failed to get balance")
+					return
+				}
+				amountMilliSats := amount * 1000
+				receiveLimit := uint64(balance.Lightning.TotalReceivable)
+				if receiveLimit < amountMilliSats {
+					logger.Logger.Info("Receive limit is less than swap in amount, ignoring")
+					return
+				}
+				onchainBalance := uint64(balance.Onchain.Spendable)
+				if onchainBalance < balanceThreshold {
+					logger.Logger.Info("Threshold requirements not met for swap, ignoring")
+					return
+				}
+				logger.Logger.WithFields(logrus.Fields{
+					"amount": amount,
+				}).Info("Initiating swap")
+				_, err = svc.SwapIn(ctx, amount, lnClient)
+				if err != nil {
+					logger.Logger.WithError(err).Error("Failed to swap")
+				}
+			case <-ctx.Done():
+				logger.Logger.Info("Stopping auto swap workflow")
+				return
+			}
+		}
+	}()
+
+	svc.swapInCancelFn = cancelFn
+
+	return nil
 }
 
-func (svc *swapsService) ReverseSwap(ctx context.Context, amount uint64, destination string, lnClient lnclient.LNClient) error {
+func (svc *swapsService) SwapOut(ctx context.Context, amount uint64, destination string, lnClient lnclient.LNClient) (string, error) {
+	if destination == "" {
+		var err error
+		destination, err = svc.cfg.Get(config.OnchainAddressKey, "")
+		if err != nil {
+			return "", fmt.Errorf("could not get onchain address from config: %s", err)
+		}
+	}
+
 	var network, err = boltz.ParseChain(svc.cfg.GetNetwork())
 	if err != nil {
-		return err
+		return "", err
 	}
 
+	// TODO: use own keys
```

Comment:
What's the benefit for using our own keys? this needs to be unique per swap right? which derivation path would we use and even if we did it would it help with recovery?